### PR TITLE
Kubevirt: fixed randomised MAC address for VM

### LINKED ISSUE #1202: Kubevirt: user cluster node not re-joining the cluster after VM pod's deletion.
Issue opened after https://support.kubermatic.com/helpdesk/tickets/3654 investigation.
Scenario on dev with StorageClass=`local-path`

- Create a KubeVirt with several nodes, for example 3.
- Wait the the 3 nodes join the user cluster.
- Delete in the infra KubeVirt Cluster one of the VM pod.

What should happen:
- a new VM pod is spawned.
- the VM should join the cluster.

In the ticket raised by the customer, the VMs is spawned on a different infra node, which might also have other issues with the storage -> to investigate (having multiple infra nodes would be required to test).
On our infra cluster, there is only 1 node. Already with the VM re-scheduled on the same infra node, the VM does not re-join the cluster.

- User cluser
```
% k get node
NAME                                          STATUS     ROLES    AGE     VERSION
awesome-bose-worker-8b9rhh-85d98758c8-2td24   Ready      <none>   179m    v1.21.8
awesome-bose-worker-8b9rhh-85d98758c8-tcxld   Ready      <none>   3h14m   v1.21.8
awesome-bose-worker-8b9rhh-85d98758c8-xl9hm   NotReady   <none>   93m     v1.21.8
```

- Infra cluster:
```
% k get po -o wide
NAME                                                              READY   STATUS    RESTARTS   AGE     IP             NODE                           NOMINATED NODE   READINESS GATES
virt-launcher-awesome-bose-worker-8b9rhh-85d98758c8-2td24-nfz8c   1/1     Running   0          3h1m    10.244.0.202   ubuntu-2004-focal-64-minimal   <none>           <none>
virt-launcher-awesome-bose-worker-8b9rhh-85d98758c8-tcxld-rws9h   1/1     Running   0          3h15m   10.244.0.198   ubuntu-2004-focal-64-minimal   <none>           <none>
virt-launcher-awesome-bose-worker-8b9rhh-85d98758c8-xl9hm-vt5xk   1/1     Running   0          72m     10.244.0.237   ubuntu-2004-focal-64-minimal   <none>           <none>
virt-launcher-infallible-feynman-worker-qjdv4t-696854f455-tjf9w   1/1     Running   0          119m    10.244.0.225   ubuntu-2004-focal-64-minimal
```

- In the KCCM logs:
```
E0304 09:36:34.861185       1 node_controller.go:354] Specified Node IP not found in cloudprovider for node "awesome-bose-worker-8b9rhh-85d98758c8-xl9hm"
```

The user cluster node:
```
 k get node awesome-bose-worker-8b9rhh-85d98758c8-xl9hm -o yaml
apiVersion: v1
kind: Node
metadata:
  annotations:
    alpha.kubernetes.io/provided-node-ip: 10.244.0.236
```

`alpha.kubernetes.io/provided-node-ip: 10.244.0.236` is the old VM pod IP, not the new one.

Need to investigate why.

### PR DESCRIPTION
**Pull request by hdurand0710**
Signed-off-by: Helene Durand <helene@kubermatic.com>

**What this PR does / why we need it**:
The PR is the first part to fix #1202. We force a fixed MAC address in the VM to have the device recognised when VM is restarted so that the system keeps the interface in the Up state.
After this PR the `Node` will be `Ready` state.
What remains to fix:
- the kubelet starts with the old ip as `--node-ip`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged)*:
Fixes #1202

**Special notes for your reviewer**:

**Optional Release Note**:

```release-note
Kubevirt: VM are created with a fixed randomised MAC address.
```

### CODE REVIEW

File: pkg/cloudprovider/util/net.go [hunk 1]
```diff
@@ -36,3 +37,19 @@ func CIDRToIPAndNetMask(ipv4 string) (string, string, int, error) {
...
 	return ip.String(), netmask, size, nil
 }
+
+// GenerateRandMAC generates a random unicast and locally administered MAC address.
+func GenerateRandMAC() net.HardwareAddr {
+	buf := make([]byte, 6)
+	var mac net.HardwareAddr
+
+	_, err := rand.Read(buf)
+	if err != nil {
```
Comment (mfranczy) [inline pkg/cloudprovider/util/net.go:47]: I guess you want to return an error here

File: pkg/cloudprovider/util/net.go [hunk 2]
```diff
@@ -36,3 +37,19 @@ func CIDRToIPAndNetMask(ipv4 string) (string, string, int, error) {
...
+	}
+
+	// Set locally administered addresses bit and reset multicast bit
```
Comment (mfranczy) [inline pkg/cloudprovider/util/net.go:50]: In the future we might want to preserve the vendor prefix, just to keep this in mind. For now it's ok.

Comment (hdurand0710) [reply pkg/cloudprovider/util/net.go:51]: ok, leave as is for now, but keeping it in mind.

File: pkg/cloudprovider/util/net.go (hunk 1 above)
Comment (hdurand0710) [reply pkg/cloudprovider/util/net.go:47]: yop! done. The whole `Create` is rejected if we can not compute the MAC address.

**Review by mfranczy — APPROVED** [review]
/approve
/lgtm

**Review by ahmedwaleedmalik — APPROVED** [review]
/approve

**Pull request merged**